modules/optimization/include/executioners/OptimizeSolve.h

```diff
@@ -0,0 +1,135 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "SolveObject.h"
+
+#include "OptimizationReporter.h"
+#include "ExecFlagEnum.h"
+#include <petsctao.h>
+#include "libmesh/petsc_vector.h"
+#include "libmesh/petsc_matrix.h"
+
+class OptimizationReporter;
+
+/**
+ * solveObject to interface with Petsc Tao
+ */
+class OptimizeSolve : public SolveObject
+{
+public:
+  static InputParameters validParams();
+  OptimizeSolve(Executioner & ex);
+
+  virtual bool solve() override;
+
+  const OptimizationReporter & getOptimizationReporter() const { return *_obj_function; }
+
+  /**
+   * Record tao data for output by a reporter
+   */
+  void getTaoSolutionStatus(std::vector<int> & tot_iters,
+                            std::vector<double> & gnorm,
+                            std::vector<int> & obj_iters,
+                            std::vector<double> & cnorm,
+                            std::vector<int> & grad_iters,
+                            std::vector<double> & xdiff,
+                            std::vector<int> & hess_iters,
+                            std::vector<double> & f,
+                            std::vector<int> & tot_solves) const;
+
+protected:
+  /// Bounds routine
+  virtual PetscErrorCode variableBounds(Tao tao);
+
+  /// Objective routine
+  virtual Real objectiveFunction();
+
+  /// Gradient routine
+  virtual void gradientFunction(libMesh::PetscVector<Number> & gradient);
+
+  /// Hessian application routine
+  virtual PetscErrorCode applyHessian(libMesh::PetscVector<Number> & s,
+                                      libMesh::PetscVector<Number> & Hs);
+
+  /// Communicator used for operations
+  const libMesh::Parallel::Communicator _my_comm;
+
+  /// List of execute flags for when to solve the system
+  const ExecFlagEnum & _solve_on;
+
+  /// objective function defining objective, gradient, and hessian
+  OptimizationReporter * _obj_function = nullptr;
+
+  /// Tao optimization object
+  Tao _tao;
+
+private:
+  bool _verbose;
+  /// tao solver info
+  int _obj_iterate = 0;
+  int _grad_iterate = 0;
+  int _hess_iterate = 0;
+  std::vector<int> _total_iterate_vec;
+  std::vector<int> _obj_iterate_vec;
+  std::vector<int> _grad_iterate_vec;
+  std::vector<int> _hess_iterate_vec;
+  std::vector<int> _function_solve_vec;
+  std::vector<double> _f_vec;
+  std::vector<double> _gnorm_vec;
+  std::vector<double> _cnorm_vec;
+  std::vector<double> _xdiff_vec;
+
+  /// Here is where we call tao and solve
+  PetscErrorCode taoSolve();
+
+  /// output optimization iteration solve data
+  static PetscErrorCode monitor(Tao tao, void * ctx);
+
+  void setTaoSolutionStatus(double f, int its, double gnorm, double cnorm, double xdiff);
+
+  ///@{
+  /// Function wrappers for tao
+  static PetscErrorCode objectiveFunctionWrapper(Tao tao, Vec x, Real * objective, void * ctx);
```

Comment:
why do we need all of these to be static?

Reply:
Yes, that's the only way we can give it to petsc while keeping it under the class namespace.